[c10d] Add hashing as a debug feature for before and after NCCL collective call

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113238

For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.

<img width="840" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/cdc70b0b-ae3c-4efd-86ff-adc5c5ba505f">


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @wz337 @tianyu-l @wconstab @yf225

fduwjj committed:
Add hashing for before and after collective call

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add hashing for before and after collective call"




[ghstack-poisoned]

fduwjj committed:
Update on "[WIP] Add hashing for before and after collective call"




[ghstack-poisoned]

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -970,6 +970,13 @@ def _is_barrier_after_init() -> int:
+    # Environment variable to control whether we hash the input/output of a collective
+    # so that we can verify that everything is working as expected at c10d level.
+    # `TORCH_NCCL_COLLECTIVE_HASH_DEBUG=1` to enable this validation.
+    return int(os.getenv("TORCH_NCCL_COLLECTIVE_HASH_DEBUG", "0"))
```
curious if we absolutely need a new env var here. The main reason against a new one is there's already a couple env vars related to distributed debugging floating around, so we should form some principles on when to add new ones or use existing ones. For example, could this be integrated into torch_distributed_debug, or is it not a good fit?

fduwjj replied:
Agree that we don't want to add too many flags here. And we want to have some principles here indeed.

For this one specifically, we don't want it to be turned on even for DEBUG because it comes with extra cost? And what if we only want to see the hash value and don't want to see other detailed logs?

fegin commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -73,6 +76,11 @@ enum ErrorHandlingMode {
 #define SHOULD_TEAR_DOWN(a) (a != NoHandling && a != CleanUpOnly)
 
+#define PRINT_COLLECTIVE_HASH_SIGNATURE(rank, opType, numel, hashValue) \
+  LOG(ERROR) << "[RANK" << rank << "] : Collective hash of " << opType  \
```
I used ERROR just to avoid any possible message missing issues. But it seems that WARN is enough. This is not critical issue though since you also use a gating variable.

fegin replied:
Not sure if we can also add the sequence number. I remember PG has a sequence number.

fduwjj replied:
yes, but this is not available in python..

fegin commented on torch/csrc/distributed/c10d/Utils.cpp:
```diff
@@ -31,4 +33,30 @@ std::vector<at::Tensor> getTensorShapes(
+    std::hash<char> hasher;
+    auto src = static_cast<const char*>(tensor.storage().data_ptr().get());
+    char* dst = (char*)std::malloc(data_size);
+    std::memset(dst, 0, data_size);
```
You can just use `calloc()`

fduwjj committed:
Update on "[WIP] Add hashing for before and after collective call"




[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/Utils.cpp:
```diff
@@ -22,4 +24,29 @@ std::vector<at::Tensor> getTensorShapes(
+    cudaMemcpy(dst, src, data_size, cudaMemcpyDeviceToHost);
+    for (size_t i = 0; i < data_size; ++i) {
+      // Update the hash for each byte in the tensor
+      hash ^= hasher(((char*)dst)[i]) + 0x9e3779b9 + (hash << 6) + (hash >> 2);
```
what hasher is used here?

fduwjj replied:
We can change the hashing algorithm. I just randomly made up one here.

wconstab replied:
well, a while back i added some hashing utils for lazy tensor that uses 128bit int https://github.com/pytorch/pytorch/blob/main/torch/csrc/lazy/core/hash.h

but there is also a hash util in c10. its probably good enough and may be faster. 
https://github.com/pytorch/pytorch/blame/main/c10/util/hash.h

might be worth comparing using one of those, but maybe yours is fine

fduwjj replied:
oh wait, I didn't see this comment. let me look into this as well.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -644,6 +644,12 @@ bool ProcessGroupNCCL::WorkNCCL::wait(std::chrono::milliseconds timeout) {
   synchronizeInternal(timeout);
   // Always return true, because abort API is not implemented.
+  if (debug_level() == DebugLevel::Detail) {
+    auto hashValue = hashTensors(*outputs_);
```
i'm confused about the terminology.  In work.wait() we are just sure that we have launched the kernel to the cuda stream, so we do not know about the tensor outputs yet right?  why not call these 'inputs'?

also, why do we have logging in both 'collective' and 'wait' - would these give different results?  (or maybe that part is intended, just to prove we didn't corrupt between creating and launching?)

In any case, the print statement coming from `PRINT_COLLECTIVE_HASH_SIGNATURE` may be confusing to users if the exact same print shows up twice for one collective.  it might be worth adding an argument for "phase" so we can print "creating" vs "launching" collective or something

fduwjj replied:
Sure, I can add stages here. But I remember @fegin mentioned that he saw discrepancies between here and inside NCCLPG.

fegin replied:
@wconstab The hash function calls `cudaMemcpy` so it will actually trigger a synchronization. `wait()` will actually wait for the collective to finish if the hash function is called. That's why it is very slow.

wconstab replied:
oh i get it now.

@fduwjj I was confused earlier, but now it makes sense why there were 2 prints, and the one in 'wait' should say "after" and the one in collective should say "before".  does this make sense?  I think the latest code only has one print, but let me review again to double check.

fduwjj committed:
Update on "[WIP][c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
Fix bugs in python code.

fduwjj committed:
Update on "[WIP][c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -4,7 +4,9 @@
 #ifdef USE_C10D_NCCL
+#include <vector>
 
+#include <cuda_runtime.h>
```
Move to NCCLUtils since we need to gate it with Macro so that builds in other environment are happy.

wconstab commented on torch/csrc/distributed/c10d/NCCLUtils.cpp:
```diff
@@ -60,6 +62,30 @@ std::string getNcclVersion() {
+        for (size_t i = 0; i < data_size; ++i) {
+          // Update the hash for each byte in the tensor
+          hash ^=
+              hasher(((char*)dst)[i]) + 0x9e3779b9 + (hash << 6) + (hash >> 2);
```
https://stackoverflow.com/questions/74805446/how-to-hash-a-pytorch-tensor

also i think @ezyang made up some kind of tensor summary that wasn't a real hash but was fast.  In this case i think its worth doing a real hash but let me see if i can find that just in case.

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -94,6 +97,12 @@ enum ErrorHandlingMode {
 #define SHOULD_TEAR_DOWN(a) (a != NoHandling && a != CleanUpOnly)
 
+#define PRINT_COLLECTIVE_HASH_SIGNATURE(phase, rank, opType, numel, hashValue) \
+  LOG(WARNING) << "[RANK" << rank << "] at phase " << phase                    \
```
update to use `logPrefix()` after rebase

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2329,6 +2336,13 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
+    auto numel = getTensorsNumel(inputs);
+    auto hashValue = hashTensors(inputs);
+    PRINT_COLLECTIVE_HASH_SIGNATURE(
+        "creating", rank_, opTypeToString(opType), numel, hashValue);
```
sorry for confusing earlier.  this one should be "before" or "input"

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -644,6 +644,12 @@ bool ProcessGroupNCCL::WorkNCCL::wait(std::chrono::milliseconds timeout) {
+    auto numel = getTensorsNumel(*outputs_);
+    auto hashValue = hashTensors(*outputs_);
+    PRINT_COLLECTIVE_HASH_SIGNATURE(
+        "launching", rank_, opTypeToString(opType_), numel, hashValue);
```
and this one should be "after" or "output" -- becuase as @fegin pointed out, `hashTensors()` will make a cuda dtoh sync and produce the real value, so it is not how i originally understood.

fduwjj committed:
Update on "[WIP][c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -94,6 +97,12 @@ enum ErrorHandlingMode {
 
+#define PRINT_COLLECTIVE_HASH_SIGNATURE(                                    \
+    prefix, phase, opType, numel, hashValue)                                \
+  LOG(WARNING) << prefix << "at phase " << phase << ": Collective hash of " \
```
could put the `logPrefix()` call in here and not require prefix arg

and simplify message to someting like 

"Hash of " << phase << " to NCCL" << opType << " with size " << numel << " was " << hashValue;
e.g.
"Hash of input to NCCL allreduce with size 256 was 0x123123123"

fduwjj replied:
> could put the `logPrefix()` call in here and not require prefix arg

Let me try.

fduwjj committed:
Update on "[c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.

<img width="840" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/cdc70b0b-ae3c-4efd-86ff-adc5c5ba505f">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.

<img width="840" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/cdc70b0b-ae3c-4efd-86ff-adc5c5ba505f">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.

<img width="840" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/cdc70b0b-ae3c-4efd-86ff-adc5c5ba505f">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.

<img width="840" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/cdc70b0b-ae3c-4efd-86ff-adc5c5ba505f">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fegin reviewed (approved):
Thanks for making the feature work!

fegin commented on torch/distributed/distributed_c10d.py:
```diff
@@ -2209,11 +2209,18 @@ def _object_to_tensor(obj, device):
 
 def _tensor_to_object(tensor, tensor_size):
+    if get_debug_level() == DebugLevel.DETAIL and is_nccl_available():
+        cuda_tensor = tensor.cuda()
```
Instead of moving to cuda(), should we just test if the tensor is on GPU? If the tensor is on CPU, it is NOT using NCCL to do the communication.

fduwjj committed:
Update on "[c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.

<img width="840" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/cdc70b0b-ae3c-4efd-86ff-adc5c5ba505f">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -648,6 +651,12 @@ bool ProcessGroupNCCL::WorkNCCL::wait(std::chrono::milliseconds timeout) {
   synchronizeInternal(timeout);
   // Always return true, because abort API is not implemented.
+  if (distDebugLevel_ >= DebugLevel::Detail) {
+    auto numel = getTensorsNumel(*outputs_);
+    auto hashValue = hashTensors(*outputs_);
```
I chatted with @fduwjj, there are thoughts to remove `outputs_` from work objects, to avoid holding output buffer's memory for longer than needed. 

I am wondering if the hashing, if needed, can be called at the user side of ProcessGroupNCCL than within ProcessGroupNCCL.

By "user side" I mean:
- called by the user at python level; or
- called within `allgather_object_list` at python level.

There, there are explicit handle to the output tensor; if any follow-up operation is performed on the tensor, extension of the tensor's lifecycle is easily understandable.

Here, `work->outputs_` is an implicit reference. It would be great if we don't have to keep it, which may lead to some surprising effects (like build-up of peak memory).

fduwjj committed:
Update on "[c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.

<img width="840" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/cdc70b0b-ae3c-4efd-86ff-adc5c5ba505f">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
@pytorchbot rebase

fduwjj committed:
Update on "[c10d] Add hashing as a debug feature for before and after NCCL collective call"


For now, we use `TORCH_DISTRIBUTED_DEBUG = DETAIL` to turn a debug feature which calculate the hashing for input tensors and output results of c10d collective in NCCL.  This is a debugging feature so that we can rule out the bug from c10d level.

<img width="840" alt="image" src="https://github.com/pytorch/pytorch/assets/6937752/cdc70b0b-ae3c-4efd-86ff-adc5c5ba505f">


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
